### PR TITLE
Votable ERC20 for interchain DAO project

### PR DESCRIPTION
**Pull request by armanthepythonguy**
* added HypERC20Votable.sol
* added delegate functionality to HypERC20Collateral.sol
* added optimizer values in hardhat config as HypERC20Votable normal size exceeds max deployment size
* added a votable option to the synthetic chain config
* changed the deploy function in ERC20Deployer to be able to deploy HypERC20Votable if the votable option is "true" in the config

### CODE REVIEW

File: contracts/HypERC20Collateral.sol [hunk 1]
```diff
@@ -15,13 +19,19 @@ contract HypERC20Collateral is TokenRouter {
     using SafeERC20 for IERC20;
 
     IERC20 public immutable wrappedToken;
+    address public tokenAddress;
```
Comment (nambrot) [inline contracts/HypERC20Collateral.sol]: Why does this need to be added? It's just the wrappedToken, no?

File: contracts/HypERC20Collateral.sol [hunk 2]
```diff
@@ -7,6 +7,10 @@ import {Message} from "./libs/Message.sol";
 import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
 
+interface delegateTokenInterface{
```
Comment (nambrot) [inline contracts/HypERC20Collateral.sol]: Can you properly vendor this interface (with a reference where it is from)

File: contracts/HypERC20.sol [hunk 3]
```diff
@@ -71,4 +71,4 @@ contract HypERC20 is ERC20Upgradeable, TokenRouter {
     ) internal override {
         _mint(_recipient, _amount);
     }
-}
+}
```
Comment (nambrot) [inline contracts/HypERC20.sol:74]: Don't check in these whitespace changes

File: contracts/HypERC20Votable.sol [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+import "@openzeppelin/contracts/token/ERC20/extensions/draft-ERC20Permit.sol";
+import "@openzeppelin/contracts/token/ERC20/extensions/ERC20Votes.sol";
+import "@openzeppelin/contracts-upgradeable/utils/ContextUpgradeable.sol";
+
+/**
+ * @title Hyperlane ERC20 Token Router that extends ERC20 with remote transfer functionality.
+ * @author Abacus Works
+ * @dev Supply on each chain is not constant but the aggregate supply across all chains is.
+ */
+contract HypERC20Votable is ERC20, TokenRouter, ERC20Permit, ERC20Votes {
```
Comment (nambrot) [inline contracts/HypERC20Votable.sol:14]: Remind me, why can't you just extend from `HypERC20` and `ERC20Votes` again?

File: hardhat.config.ts [hunk 5]
```diff
@@ -12,6 +12,12 @@ module.exports = {
     compilers: [
       {
         version: '0.8.16',
+        settings: {
```
Comment (nambrot) [inline hardhat.config.ts:15]: Why is this neccessary?

File: contracts/extensions/HypERC721URIStorage.sol [hunk 6]
```diff
@@ -64,7 +64,7 @@ contract HypERC721URIStorage is HypERC721, ERC721URIStorageUpgradeable {
     function supportsInterface(bytes4 interfaceId)
         public
         view
-        override(ERC721EnumerableUpgradeable, ERC721Upgradeable)
+        override(ERC721EnumerableUpgradeable, ERC721URIStorageUpgradeable)
```
Comment (nambrot) [inline contracts/extensions/HypERC721URIStorage.sol:67]: Seems unrelated?

File: src/config.ts [hunk 7]
```diff
@@ -58,10 +59,12 @@ export const isUriConfig = (config: TokenConfig) =>
   config.type === TokenType.collateralUri;
 
 export type HypERC20Config = GasRouterConfig & SyntheticConfig & ERC20Metadata;
+export type HypERC20VotableConfig = GasRouterConfig & SyntheticConfig & ERC20Metadata;
```
Comment (nambrot) [inline src/config.ts]: Why is this neccessary?

File: src/deploy.ts [hunk 8]
```diff
@@ -123,9 +124,9 @@ export class HypERC20Deployer extends GasRouterDeployer<
   ): Promise<HypERC20> {
     const router = await this.deployContractFromFactory(
       chain,
-      new HypERC20__factory(),
-      'HypERC20',
-      [config.decimals],
+      config.votable ? new HypERC20Votable__factory : new HypERC20__factory(),
+      config.votable ? 'HypERC20Votable' : "HypERC20",
+      config.votable ? [] : [config.decimals],
```
Comment (nambrot) [inline src/deploy.ts:130]: If `HypERC20Votable` is just inherting from `HypERC20`, then they should have the same args?

File: contracts/HypERC20Collateral.sol [hunk 9]
```diff
@@ -15,13 +19,19 @@ contract HypERC20Collateral is TokenRouter {
...
     /**
      * @notice Constructor
      * @param erc20 Address of the token to keep as collateral
      */
     constructor(address erc20) {
         wrappedToken = IERC20(erc20);
+        tokenAddress = erc20;
+    }
+
+    function delegateVotes(address _L1VoteDelegator) external onlyOwner{
```
Comment (nambrot) [inline contracts/HypERC20Collateral.sol]: Actually thinking about this more, the typical approach to implement ERC20 supersets is to create another version, so I think you should do the same think here as with HypERC20 and create a `HypERC20CollateralVotes` contract that inherits from `HypERC20Collateral`

File: contracts/HypERC20Collateral.sol (hunk 1 above)
Comment (armanthepythonguy) [reply contracts/HypERC20Collateral.sol]: IERC20 doesn't have the delegate function which is required to delegate the voting power to the L1 Vote Delegator.

File: hardhat.config.ts (hunk 5 above)
Comment (armanthepythonguy) [reply hardhat.config.ts:15]: The hypERC20Votable contract size exceeds the maximum contract size. Thus we need an optimizer in order to deploy it.

File: src/deploy.ts (hunk 8 above)
Comment (armanthepythonguy) [reply src/deploy.ts:130]: HypERC20Votable is inheriting the ERC20 contract which already has fixed decimals. So no need to initialize it as we do in normal HypERC20.

File: contracts/HypERC20Collateral.sol (hunk 9 above)
Comment (armanthepythonguy) [reply contracts/HypERC20Collateral.sol]: Got it will create a new HypERC20CollateralVotes

File: contracts/HypERC20Votable.sol (hunk 4 above)
Comment (armanthepythonguy) [reply contracts/HypERC20Votable.sol:14]: The HypERC20 contract uses ERC20Upgradeable. But the ERC20Votes extension needed for voting uses ERC20. Thus there will be conflicting declarations.

File: contracts/extensions/HypERC721URIStorage.sol (hunk 6 above)
Comment (armanthepythonguy) [reply contracts/extensions/HypERC721URIStorage.sol:67]: I am not able to compile without this change. Following is the error I am getting :- 

TypeError: Function needs to specify overridden contract "ERC721URIStorageUpgradeable".
  --> contracts/extensions/HypERC721URIStorage.sol:67:9:
   |
67 |         override(ERC721EnumerableUpgradeable, ERC721Upgradeable)
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Note: This contract: 
  --> @openzeppelin/contracts-upgradeable/token/ERC721/extensions/ERC721URIStorageUpgradeable.sol:13:1:
   |
13 | abstract contract ERC721URIStorageUpgradeable is Initializable, IERC4906Upgradeable, ERC721Upgradeable {
   | ^ (Relevant source part starts here and spans across multiple lines).


TypeError: Invalid contract specified in override list: "ERC721Upgradeable".
  --> contracts/extensions/HypERC721URIStorage.sol:67:9:
   |
67 |         override(ERC721EnumerableUpgradeable, ERC721Upgradeable)
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Note: This contract: 
  --> @openzeppelin/contracts-upgradeable/token/ERC721/ERC721Upgradeable.sol:20:1:
   |
20 | contract ERC721Upgradeable is Initializable, ContextUpgradeable, ERC165Upgradeable, IERC721Upgradeable, IERC721MetadataUpgradeable {
   | ^ (Relevant source part starts here and spans across multiple lines).


Error HH600: Compilation failed

File: src/config.ts (hunk 7 above)
Comment (armanthepythonguy) [reply src/config.ts]: I have removed it in the new update

File: contracts/HypERC20CollateralVotable.sol [hunk 10]
```diff
@@ -0,0 +1,22 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity >=0.8.0;
+
+import "./HypERC20Collateral.sol";
+
+interface IL1Token{
```
Comment (nambrot) [inline contracts/HypERC20CollateralVotable.sol:6]: Still properly vendor this interface

File: contracts/HypERC20CollateralVotable.sol [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    function delegate(address delegatee) external;
+}
+
+
+contract HypERC20CollateralVotable is HypERC20Collateral{
+
+    IL1Token L1Token;
```
Comment (nambrot) [inline contracts/HypERC20CollateralVotable.sol:13]: Is this neccessary? Imo you can just do this inside `delegateVotes` no?

File: contracts/HypERC20Votable.sol (hunk 4 above)
Comment (nambrot) [reply contracts/HypERC20Votable.sol:14]: Can you not use https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable/blob/master/contracts/token/ERC20/extensions/ERC20VotesUpgradeable.sol

File: contracts/extensions/HypERC721URIStorage.sol (hunk 6 above)
Comment (nambrot) [reply contracts/extensions/HypERC721URIStorage.sol:67]: Ah yes i see
